keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,558 @@
+
+<<[/UNRESOLVED]>>
+
+
+```
+
+* define another object (ie. IPAddress and NodePort) that represent the allocated resources, the name of
+the object will be the port or the IP, so we can guarantee the consistency.
+
+* the allocated object must reference the object used to define the range allocation, and validated against it (i.e. IP within range)
+
+```
+<<[UNRESOLVED allocation mode]>>
+
+* and allocated object can have different status: allocated, reserved, free, ...
+
+* on service creation, if the NodePort or ClusterIP field is set:
+1. check that the object doesn't exist in the apiserver and Create a new one (owned by the Service)
+2. check if the object exists and the Status is Free and change the status.
+
+* on service creation, if the NodePort or ClusterIP field is not set, allocate a random one from the defined range in the corresponding Object (IPRange or NodePortRange)
+
+There can be different model to allocate a random resource:
+
+1. Same as today, the apiserver can keep an index with current status and start with a random offset to
+find a free resources.
+
+2. A controller create a pool of free resource Objects and the apiserver picks one of them, this adds 
+more complexity to the model, but will improve the time to allocate a new free resources since is just
+popping an object from a list.
+
+<<[/UNRESOLVED]>>
+```
+
+## Design Details
+
+
+### Cluster IPs
+
+#### IP Ranges
+
+Define two new objects, a new IPRange Object that defines the range used to allocate Services Cluster IPs
+
+```go
+// IPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type IPRange struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPRangeSpec   `json:"spec,omitempty"`
+	Status IPRangeStatus `json:"status,omitempty"`
+}
+
+// IPRangeSpec describe how the IPRange's specification looks like.
+type IPRangeSpec struct {
+	// Range of IPs in CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+	Range string `json:"range"`
+	// Primary indicates if this is the primary allocator to be used by the
+	// apiserver to allocate IP addresses.
+	// NOTE this can simplify the Service strategy logic so we don't have to infer
+	// the primary allocator, it also may allow to switch between primary families in
+	// a cluster, but this looks like a loooong shot.
+	// +optional
+	Primary bool `json:"primary,omitempty"`
+}
+
+// IPRangeStatus defines the observed state of IPRange.
+type IPRangeStatus struct {
+	// Free represent the number of IP addresses that are not allocated in the Range.
+	// +optional
+	Free int64 `json:"free,omitempty"`
+}
+```
+
+Reserve well-known-names for the IPRanges objects, that represent current allocators service-subnet-ranges:
+
+```go
+// default allocator well-known-names
+const (
+	DefaultIPv4Allocator = "allocator.ipv4.k8s.io"
```

Comment:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2593-multiple-cluster-cidrs decided on making one "pool" have (optionally) both v4 and v6.  I think the assumption was that sometimes (e.g. converting from flaags in this case) you want to program them together.

@rahulkjoshi @mskrocki - is that still correct?

If we pursue this model, they feel similar enough that the APIs should probably be similar.

In this case it means you can simply call the converted instance "default"

Reply:
We decided on having both v4 and v6 grouped together to avoid race conditions. This explicitly identifies the mode given node has to be in: single or dualstack.

Reply:
yeah, I remember that discussion, and grouping was the conclusion